docs/combinatorics.rst

```diff
@@ -1,6 +1,202 @@
+relating to tree ranks used in this process.
+
+***********
+Definitions
+***********
+
+shape rank
+    For a tree :math:`T` with :math:`n` leaves, the shape rank is the position
+    of :math:`T`'s shape in the enumeration of all tree shapes with
+    :math:`n` leaves.
+
+label rank
+    For a tree :math:`T`, the label rank is the position of the labelling of
+    :math:`T`'s leaves in the enumeration of all unique ways the leaves of
+    :math:`T` could be labelled.
+
+rank
+    The rank of a leaf-labelled tree is considered to be the tuple of its
+    ``shape rank`` and ``label rank``.
+
+unrank
+    The action of reconstructing a leaf-labelled tree from its ``rank``.
+
+++++++++++++++++++++++
+Enumerating Topologies
+++++++++++++++++++++++
+
+Below we outline the approach for enumerating all unique topologies
+of leaf-labelled trees with :math:`n` leaves. This takes part in two
+steps. First, we must consider how to generate unique tree shapes.
+Then, for each shape we must consider how to uniquely assign labels
+to the leaves. Note that this enumeration is not limited to binary trees.
+
+***********************
+Enumerating Tree Shapes
+***********************
+Let :math:`S_n` denote the set of all unlabelled trees, or tree shapes
+of :math:`n` leaves.
+We can see that :math:`S_1` has one element: a single leaf.
+
+A tree with :math:`n > 1` leaves can be obtained by joining at least two trees whose number
+of leaves sum to :math:`n`. This maps very closely to the concept of integer partitions.
+We first compute all integer partitions* of `n`, and then for each partition :math:`P`,
+compute the cross product of all :math:`k`-leaf trees for all :math:`k \in P`.
+For example, for :math:`n = 5` and :math:`P = [2, 3]`, we can see that
+:math:`S_2 \times S_3` produces valid trees in :math:`S_5`.
+
+Below are all tree shapes in :math:`S_4`, along with the partitions of 4 that
+generate them.
+
+.. image:: _static/four_leaf_tree_shapes.png
+   :alt: All four-leaf tree shapes and their generating partitions
+
+Note that the middle column reflects all trees in :math:`S_3` in the right subtree.
+
+`*` This excludes the partition [:math:`n`], since this would create a unary node
+and trees with unary nodes are inumerable (and potentially infinite).
+
+.. note::
+    Using non-decreasing integer partitions enforces a *canonical orientation* on the tree
+    shapes, where children under a node are ordered by the number of leaves below them.
+    This is important because it prevents us from repeating trees that are topologically
+    the same but whose children are ordered differently.
+
+However, this method has the potential to double-count trees when we encounter
+partitions with repeated numbers of leaves. To remove these redundant trees we
+reduce the set of trees so that if a subtree has the same number of leaves as
+its left sibling, it must have an equal or greater shape rank.
```

Comment:
very nice solution!

Reply:
So, I guess that what you're doing is saying the subtrees must be ordered by `(size, rank)`? So the rank of a tree among trees with `n` leaves is the same as the rank of its ordered list of subtrees: `((size_1, rank_1), ..., (size_n, rank_n))` among all such ordered lists?

Reply:
Yes, for unlabelled trees. The definition is extended when you add labels. Precisely, it's

https://github.com/tskit-dev/tskit/blob/72ac9d2472f645163c0675c4413ddc3cb6e709b6/python/tskit/combinatorics.py#L264-L269